src/tools/run-make-support/src/lib.rs

```diff
@@ -73,11 +73,43 @@ pub fn static_lib_name(name: &str) -> String {
     //     endif
     // endif
     // ```
-    assert!(!name.contains(char::is_whitespace), "name cannot contain whitespace");
+    assert!(!name.contains(char::is_whitespace), "static library name cannot contain whitespace");
 
     if target().contains("msvc") { format!("{name}.lib") } else { format!("lib{name}.a") }
 }
 
+/// Construct a path to a dynamic library under `$TMPDIR` given the library name. This will return a
+/// path with `$TMPDIR` joined with platform-and-compiler-specific library name.
+pub fn dynamic_lib(name: &str) -> PathBuf {
+    tmp_dir().join(dynamic_lib_name(name))
+}
+
+/// Construct the dynamic library name based on the platform.
+pub fn dynamic_lib_name(name: &str) -> String {
+    // See tools.mk (irrelevant lines omitted):
+    //
+    // ```makefile
+    // ifeq ($(UNAME),Darwin)
+    //     DYLIB = $(TMPDIR)/lib$(1).dylib
+    // else
+    //     ifdef IS_WINDOWS
+    //         DYLIB = $(TMPDIR)/$(1).dll
+    //     else
+    //         DYLIB = $(TMPDIR)/lib$(1).so
+    //     endif
+    // endif
+    // ```
+    assert!(!name.contains(char::is_whitespace), "dynamic library name cannot contain whitespace");
+
+    if target().contains("darwin") {
+        format!("lib{name}.dylib")
+    } else if target().contains("windows") {
```

Comment:
We have a bunch of `target.contains("windows")` etc. in bootstrap, and it's kind of annoying to figure out what string should be used to match some known OSes. It would be nice to introduce some helper functions/methods like `is_windows`, `is_macos` etc., to make this easier (doesn't have to be in this PR).

Reply:
There actually already are some functions like this in this same file:

```rs
/// Check if target is windows-like.
pub fn is_windows() -> bool {
    env::var_os("IS_WINDOWS").is_some()
}

/// Check if target uses msvc.
pub fn is_msvc() -> bool {
    env::var_os("IS_MSVC").is_some()
}
```

However, they were not used in the `static_library_name` function, even though [both of these were added in the same commit by Jieyou](https://github.com/rust-lang/rust/commit/142d02d472246c25ecfa9d7d767c3a97785f010f). I suspected there was some subtle reason to use the `target().contains` syntax, so I reproduced it in this PR.

Reply:
This is inconsistent because I was being inconsistent with myself (heh). Would be nice to have a follow-up PR to use add helper methods like `is_windows()`, `is_darwin()` etc. that more consistently matches on `target()` (which uses `TARGET` env var), or use `uname` for determining the specific distro as needed.